### PR TITLE
Divide and Conquer Algorithm

### LINKED ISSUE #384: [Feature Request] Divide and Conquer
🚀 Feature Request: Divide and Conquer Implementation in Algorithm Optimization

 Is your feature request related to a problem? Please describe.
"I'm frustrated when I need to solve large-scale problems, and the current solutions take too long to execute or are inefficient. For example, sorting large datasets or finding specific elements in large arrays becomes slow using traditional iterative methods like Bubble Sort or linear search."
Describe the solution you'd like-
I would like to implement Divide and Conquer as a solution to improve the efficiency of problem-solving. By breaking down the problem into smaller subproblems, solving them independently, and combining their solutions, I can achieve faster algorithms. Examples include using Merge Sort for sorting, Quick Sort for partitioning datasets, and Binary Search for finding elements in a sorted array.

 Describe alternatives you've considered
Alternatives include:
- Iterative approaches like Bubble Sort or Selection Sort for sorting problems, but these tend to have a time complexity of O(n²), making them inefficient for large inputs.
- Brute-force methods like linear search for finding elements in an array, but they take O(n) time for unsorted arrays, which can be slow for large datasets.
  
None of these alternatives provide the efficiency gains that Divide and Conquer offers with O(n log n) for sorting and O(log n) for searching.
 Additional context-
The Divide and Conquer approach is applicable in various algorithmic challenges, including sorting (Merge Sort, Quick Sort), searching (Binary Search), matrix multiplication (Strassen’s Algorithm), and computational geometry (Convex Hull). It provides a general framework for improving time complexity by recursively breaking down problems into smaller, more manageable subproblems.

By adopting this method, I can significantly reduce the computational cost of problems that grow exponentially with size.

**Would you like to work on this feature?**  
- [x] Yes  
- [ ] No  

assign me this task to work and tag with hactoberfest-accepted and gssoc-ext

### PR DESCRIPTION
**Pull request by pragna7**
Description
New Algorithm-Divide and Conquer Algorithm

Fixes #384

### Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update

### Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Review by ajay-dhangar — APPROVED** [review]
ok

**Comment by pragna7** [comment]
could you please review it and merge it

**Comment by pragna7** [comment]
Can I know the issue occurred , what changes i have to make?

File: docs/algorithms/Divide and Conquer Algorithm/Strassen'sAlgorithm.md [hunk 1]
```diff
@@ -0,0 +1,165 @@
+---
```
Comment (ajay-dhangar) [inline docs/algorithms/Divide and Conquer Algorithm/Strassen'sAlgorithm.md:1]: first you need rename your file `Strassen'sAlgorithm.md` to `strassen-algorithm.md` for best practices

File: docs/algorithms/Divide and Conquer Algorithm/Strassen'sAlgorithm.md [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+  - Combine the computed products to form the resultant matrix using additions and subtractions.
+
+- **Reduced Multiplications**:
+  - Strassen's algorithm performs only 7 multiplications for two \(2 \times 2\) matrices instead of the 8 required in the standard approach.
+
+### Time Complexity:
+
+- **Time Complexity: O(n^{\log_2 7}) ≈ O(n^{2.81})**  
+  Strassen's algorithm reduces the complexity of multiplying two \(n \times n\) matrices from \(O(n^3)\) to approximately \(O(n^{2.81})\).
+
```
Comment (ajay-dhangar) [inline docs/algorithms/Divide and Conquer Algorithm/Strassen'sAlgorithm.md:32]: replace your some code like

1. `O(n^{\log_2 7}) ≈ O(n^{2.81})` to `$O(n^{\log_2 7}) ≈ O(n^{2.81})$`
2. `\(n \times n\)` to `$n \times n$`
3. `\(O(n^3)\)` with `$O(n^3)$`
4. `\(O(n^{2.81})\)` with `$O(n^{2.81})$`

File: docs/algorithms/Divide and Conquer Algorithm/Strassen'sAlgorithm.md [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+### Explanation of the Code:
+
+1. **Matrix Addition and Subtraction**:
+   - The `add` function takes two matrices and returns their sum.
+   - The `subtract` function takes two matrices and returns their difference.
+
+2. **Strassen's Algorithm Function**:
+   - The `strassen` function performs Strassen's matrix multiplication.
+   - If the input size is \(1 \times 1\), it multiplies the single elements.
```
Comment (ajay-dhangar) [inline docs/algorithms/Divide and Conquer Algorithm/Strassen'sAlgorithm.md:153]: replace `\(1 \times 1\)` with `$1 \times 1$`

File: docs/algorithms/Divide and Conquer Algorithm/Strassen'sAlgorithm.md [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+   - For larger matrices, it divides them into four submatrices (quadrants).
+   - It calculates the seven products using recursive calls to the `strassen` function based on the Strassen formulas.
+   - Finally, it combines the results to form the resultant matrix.
+
+3. **Driver Code**:
+   - The `main` function initializes two \(2 \times 2\) matrices and calls the `strassen` function to multiply them.
```
Comment (ajay-dhangar) [inline docs/algorithms/Divide and Conquer Algorithm/Strassen'sAlgorithm.md:159]: replace `\(2 \times 2\)` with `$2 \times 2$`

File: docs/algorithms/Divide and Conquer Algorithm/Strassen'sAlgorithm.md [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+   - It prints the resultant matrix.
+
+### Summary:
+
+Strassen's Algorithm significantly reduces the complexity of matrix multiplication by using a divide-and-conquer approach, achieving a time complexity of approximately \(O(n^{2.81})\). Although it is more efficient than the naive approach, it is more complex to implement and may not be optimal for smaller matrices due to the overhead of recursive calls. Strassen's algorithm is particularly useful in scenarios where large matrix multiplications are needed, such as in scientific computing and graphics applications.
```
Comment (ajay-dhangar) [inline docs/algorithms/Divide and Conquer Algorithm/Strassen'sAlgorithm.md:164]: replace `\(O(n^{2.81})\)` with `$O(n^{2.81})$`

**Review by ajay-dhangar — CHANGES_REQUESTED** [review]
update your code based on suggestions and also, applied in all file and read more about [Math Equations](https://docusaurus.io/docs/markdown-features/math-equations)

**Comment by pragna7** [comment]
i committed changes can you merge it now

File: docs/algorithms/Divide and Conquer Algorithm/Strassen'sAlgorithm.md (hunk 5 above)
Comment (pragna7) [reply docs/algorithms/Divide and Conquer Algorithm/Strassen'sAlgorithm.md:164]: i committed changes

**Pull request closed without merging**